SpringSecurityの認可認証を実装、テストコードを作成しました。レビューよろしくお願いします。

RYA234 opened the pull request:
認可、認証を実装しました。
パスワードのBcrypt化の確認テストコードがシンプルになっています。
expected != actualであればOKという考えで実装しています。
これでも問題ないでしょうか？

RYA234 committed:
テストコード実装

RYA234 committed:
ログアウト機能を実装

RYA234 committed:
テストコード実装

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Test
+    @WithMockUser
+    @DisplayName("ログイン後に認証できるか確認")
```
@RYA234 

displayNameがおかしいです。

認証（誰であるか）するためにログインするのですよね？
テスト名から何を検証するものであるか明示するとよいと思います。

認証後に**エンドポイントのリクエストすると、status code 200 が返る
などするかもしれません

そしてpermitall()は認証しなくてもアクセスできるエンドポイントではありませんか？

https://spring.pleiades.io/spring-security/site/docs/5.3.13.RELEASE/reference/html5/#servlet-authorization-filtersecurityinterceptor

RYA234 replied:
displayNameがおかしいです。
→修正しました。

そしてpermitall()は認証しなくてもアクセスできるエンドポイントではありませんか？
→はい。テストの項目がおかしいですね。
実装するコードは
        mvc.perform(get("/check/not_access"))
        .andExpect(status().isOk());
        だけで十分ですね

Umizoko replied:
認証後にアクセスできるかどうか確認するには
エンドポイント一つだけでいいと思います。

Get /check/not_access は別の意図に使われそうです。
このエンドポイントはログイン成功後にリダイレクトするURLですかね。

認証後にアクセスできるエンドポイントを別途作成するのがよいと思います
アカウントページへのエンドポイントとかでしょうか

RYA234 replied:
エンドポイント作成しました。
menu/menuです。

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Test
+    @WithMockUser
+    @DisplayName("ログアウトができるか確認")
```
IMO

認証後にログアウトのエンドポイントへリクエストすると、~へリダイレクトする

など書くかもしれません。

assertがないので、何を検証しているのかわかりません

RYA234 replied:
修正しました。
Assertで比較したので検証対象がわかりやすくなったと考えます。
以下のURLを参考にしました。　ログアウトの項目です。
https://qiita.com/opengl-8080/items/eaa8f4eb9286a3df7986#%E3%83%AD%E3%82%B0%E3%82%A2%E3%82%A6%E3%83%88

Umizoko replied:
MockMvcにアサーションの機能があるので
それでlocationヘッダーの値の比較を行うとよいと思います

```java
    @Test
    @WithMockUser
    @DisplayName("認証後にログアウトのエンドポイントへリクエストすると、/logoutへリダイレクトする")
    public void Logout() throws Exception {
        mvc.perform(logout())
                .andExpect(status().is3xxRedirection())
                .andExpect(header().string("Location", "http://localhost:8080/logout?logout"));
    }
```

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Test
+    @DisplayName("未ログイン時のアクセスチェック")
```
IMO

認証前に**のエンドポイントへリクエストすると、xxへリダイレクトする

とかくかもしれないです。

リダイレクトはlocation headerへ飛ぶので、location headerのアサーションするといいかもですね

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,83 @@
+    @WithMockUser
+    @DisplayName("パスワードがBcryptCheckされているか確認")
+    public void passwordBCryptCheck(){
+         String  unexpected = "PASS"; 
```
インデント修正しましょう

RYA234 replied:
修正しました。

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Test
+    @WithMockUser
```
WithMockUserアノテーション必要でしょうか

RYA234 replied:
必要です。
WithMockUserが無いとログインした状態になりません。

Umizoko replied:
ログイン必要でしょうか
文字列が正しくエンコードされているかのみのテストのみですよね

RYA234 replied:
必要ないですね。勘違いしていました。修正します。お手数掛けました。

Umizoko commented on src/main/java/com/example/spring_boot_memo/SpringSecurity/config/SecurityConfig.java:
```diff
@@ -0,0 +1,57 @@
+@Configuration
+public class SecurityConfig  extends WebSecurityConfigurerAdapter {
+
+    @Bean
```
Q

bean定義していますが、参考にしているものありますか？
４８行目で
```
PasswordEncoder encoder = passwordEncoder();
```

このように定義しているのは使い方あっているのでしょうか？

RYA234 replied:
下記URLのtestBcryptメソッドを参考にしました。
実装　SpringSecurity設定の項目です。

https://qiita.com/ryo0527/items/dbdbb2a476cc74e9116b

Umizoko replied:
公式ドキュメントはありますか？

Beanアノテーション登録したものはAutowiredでインジェクトさせるのがセオリーな気がするのですが

いい記事見つからないですが、こちらが実装の参考になると思います
https://www.baeldung.com/spring-security-registration-password-encoding-bcrypt

RYA234 replied:
新しいConfigファイル(BeanConfig.java)を作成して、Beanを作成しました。

Umizoko commented on src/main/java/com/example/spring_boot_memo/SpringSecurity/config/SecurityConfig.java:
```diff
@@ -0,0 +1,57 @@
+        auth
+                .inMemoryAuthentication()
+                .withUser("User") // userを追加
+                .password(encoder.encode("PASS"))
```
q

パスワードに固定値が入ってますが、大丈夫でしょうか
参考にした公式ドキュメントありますか

RYA234 replied:
公式ドキュメント
https://spring.pleiades.io/spring-security/site/docs/5.1.7.RELEASE/reference/html/jc.html#jc-authentication-inmemory

インメモリー用だから直打ちになっていますね。
デプロイするときはコメントアウトしてると予想？

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Test
+    @WithMockUser
+    @DisplayName("パスワードがBcryptCheckされているか確認")
```
imo

エンコードした文字列の比較の方がよいかもですね

RYA234 replied:
bcryptは文字列の比較をするには手間がかかります。
ソルト（ランダムな文字列）を使っているため単純な比較ができません。
bcryptのアルゴリズムレベルでテストコードを時間かかると思います。（3hぐらい）
本格的なテストコード書きますか？
とりあえず以下のURLに沿って書いてみたのですが、難しいですね。
ストレッチング回数がどこにかかっているかわかりません。
自分の考えとしては、


１．パスワードを１０００回ハッシュ化する。  
２．１とソルトの値を足してハッシュ化する。  
３．<br>



https://medium-company.com/bcrypt/

RYA234 replied:
インターフェース PasswordEncoderのmatchesメソッド使ったら簡単に実装できますね。
エンコード後の生のパスワードとストレージからのエンコードされたパスワードを比較してるみたいですね。

公式ドキュメント
https://spring.pleiades.io/spring-security/site/docs/current/api/org/springframework/security/crypto/password/PasswordEncoder.html

RYA234 committed:
コメント追加

RYA234 committed:
プルリクエスト修正しました。

RYA234 committed:
passwordBCryptMatchCheckMethod大幅修正

RYA234 committed:
プルリクエストの内容更新

whenGetCustomers_thenStatus200()
遷移先画面の変更に伴って変更

AccessCheckInNoAuth()について
アクセスできないviewのリダイレクト先がわかるように修正

passwordBCryptMatchCheck()の@WithMockUser削除
PasswordEncoderの呼び出しは@Autowiredを使用

RYA234 committed:
新規作成
ログイン後の画面

RYA234 committed:
ログインの画面とPasswordEncoderを@Autowired化

RYA234 committed:
PasswordEncoderをbean化

RYA234 committed:
menuのコントローラーを追加

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,74 @@
+    @WithMockUser
+    @DisplayName("認証後にログアウトのエンドポイントへリクエストすると、/logoutへリダイレクトする")
+    public void Logout() throws Exception {
+        mvc.perform(logout())
```
[IMO]
ステータスコードのアサーションも行うといいと思います
リダイレクトのコードですね

RYA234 replied:
すいません。わからないところが多いのでlogoutのテストコードが実装がままならないです。
通話で確認する機会が欲しいです。

Umizoko commented on src/main/java/com/example/spring_boot_memo/SpringSecurity/config/SecurityConfig.java:
```diff
@@ -0,0 +1,58 @@
+    }
+    @Override
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception{
+        // PasswordEncoder encoder = passwordEncoder();
```
[MUST]

コメントアウト（不要なコード）は削除してください

RYA234 replied:
削除しました。

Umizoko commented on src/main/java/com/example/spring_boot_memo/SpringSecurity/Controller.java:
```diff
@@ -0,0 +1,40 @@
+        return "check/OK";
+    }
+
+    // SecurityConfigで/logout→/check/not_accessに遷移する
```
[IMO]

```
// SecurityConfigで/logout→/check/not_accessに遷移する
```

このコメントは正しいですか。正しくなのであれば削除してください。
ログイン後の認証成功・失敗のリダイレクト先のＵＲＬでしょうか

コメントには注意を払ってください。ＤＲＹの原則はコメントにも適応されます。
コードとコメントで意図が分かれてしまうのであれば、コメント側は削除するようにするといいと思います

RYA234 replied:
削除しました。

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Test
+    @WithMockUser
+    @DisplayName("認証後にログアウトのエンドポイントへリクエストすると、/logoutへリダイレクトする")
```
```suggestion
    @DisplayName("認証後にログアウトのエンドポイントへリクエストすると、/logout?logoutへリダイレクトする")
```

LocationのURL、パスしか表現していないように見えます。
ドメイン含めたＵＲＬにする必要はありませんか？

環境の場合
https://{your-domain}/logout?logout

Umizoko replied:
tesuto

RYA234 replied:
ドメインを含む方法があるか調べてみてたのですが、見つからなかったです。
実装方法がわからん or 時間がかかりそうです。

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,74 @@
+        mvc.perform(get("/check/OK"))
+                .andExpect(header().string("Location","http://localhost/login"));        // アクセスできないviewはloginへリダイレクト
+        mvc.perform(get("/check/OK"))
+                .andExpect(header().string("Location","http://localhost/login"));        // アクセスできないviewはloginへリダイレクト
```
[MUST]

コードが重複しているように見えます。
重複しているのであれば、一方は削除してください

RYA234 replied:
削除しました。

Umizoko commented on src/test/java/com/example/spring_boot_memo/SecurityTest.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Test
+    @DisplayName("認証成功前のアクセス状況のチェック")
```
[IMO]

何に対するテストですか？
意図を明確にしてください

認証が必要なAPIに、認証されてないリクエストのとき、loginページにリダイレクトすることを検証したいテストであれば、それをDisplayNameに記述するとより分かりやすくなると思います

そうすると
```
// アクセスできないviewはloginへリダイレク
```
このコメントを削除しても意図がわかるテストになると思います

Umizoko commented on src/main/java/com/example/spring_boot_memo/SpringSecurity/config/SecurityConfig.java:
```diff
@@ -0,0 +1,58 @@
+        http.csrf().disable();
+    }
+    @Override
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception{
```
[IMO]

私が知らないので教えてほしいのです。
これは何をしている関数でしょうか

https://spring.io/guides/topicals/spring-security-architecture

RYA234 replied:
ログイン認証に使用するユーザー情報を設定する関数です。
DB はまだ使わないので、inMemoryAuthentication というメソッドを使用して、メモリ内にユーザー情報を格納して認証を行うようにしています。
3-1-3参照
https://qiita.com/t-yama-3/items/a538d47b8f0a27639d23#3-1-3-%E3%83%A6%E3%83%BC%E3%82%B6%E3%83%BC%E6%83%85%E5%A0%B1%E3%82%92%E3%82%B3%E3%83%BC%E3%83%89%E3%81%AB%E8%A8%98%E8%BC%89

Umizoko reviewed (commented):
@RYA234 
気になるところコメントしました
確認おねがいします

RYA234 committed:
プルリクエストの内容更新２
